R/kmeans-clustering-method.R

```diff
@@ -0,0 +1,528 @@
+      # K-Means
+      private$kmeans.fit <- kmeans(df.stand, numOfClusters)
+      # Save the centers 
+      private$centers <- private$kmeans.fit[["centers"]]
+      private$cluster <- private$kmeans.fit[["cluster"]]
+      #print(kmeans.fit)
+      #return(invisible(private$kmeans.fit))
+    },
+    
+    # Generate confusion matrix
+    calculateConfusion = function() {
+      # generate a confusion matrix of clusters and labels
+      k.means.fit <- self$getKmeansfit()
+      clusters <- k.means.fit$cluster
+      #labels <- super$getLabelColVal()
+      labels <- private$labelColValues
+      numOfLabels <- length(unique(labels)) - sum(is.na(unique(labels)))
+      d <- data.frame(state = labels, cluster = clusters)
+      td <- as.data.frame(table(d))
+      # convert from raw counts to percentage of each label
+      private$confusionMatrix <- matrix(ncol = max(clusters), nrow = 0) # k col
+      for (i in 1:numOfLabels) {
+        total <- sum(td[td$state == td$state[i],3])
+        private$confusionMatrix <- rbind(private$confusionMatrix, td[td$state == td$state[i],3]/total)
+      }
+      rownames(private$confusionMatrix) <- td[1:numOfLabels,1]
+      return(invisible(private$confusionMatrix))
+    },
+    
+    # Assign labels to the clusters
+    assignClusterLabels = function() {
+      cm <- self$getConfusionMatrix()
+      k <- nrow(self$getKmeansfit()[["centers"]])
+      # take the cluster label from the highest percentage in that column
+      private$cluster.labels <- list()
+      for (i in 1:k) {
+        private$cluster.labels <- rbind(private$cluster.labels, 
+                                         row.names(cm)[match(max(cm[,i]), cm[,i])])
+      }
+      
+      # this may still miss some labels, so make sure all labels are included
+      for (l in rownames(cm)) { 
+        if (!(l %in%  private$cluster.labels)) 
+        { 
+          cluster.number <- match(max(cm[l,]), cm[l,])
+          private$cluster.labels[[cluster.number]] <- c(private$cluster.labels[[cluster.number]], l)
+        } 
+      }
+    },
+    
+    ## Generate the data frame that Combine grain.col, cluster labels, 
+    ## and time to be put back into SAM table
+    createDf = function() {
+      dtStamp <- as.POSIXlt(Sys.time())
+
+      private$outDf <- data.frame(
+        0,                                 # BindingID
+        'R',                               # BindingNM
+        dtStamp,                           # LastLoadDTS
+        private$grainColValues,            # GrainID
+        private$labelColValues,            # labelCol
+        private$cluster)                   
+      
+      
+      colnames(private$outDf) <- c(
+        "BindingID",
+        "BindingNM",
+        "LastLoadDTS",
+        self$params$grainCol,
+        self$params$labelCol,
+        "cluster"
+      )
+      
+      # Remove columns that are only NA
+      private$outDf <- private$outDf[,colSums(is.na(private$outDf)) < nrow(private$outDf)]
```

Comment:
Do you mind breaking this out into a function in common-column-row-expectations.R and adding unit tests around it?

Would be super helpful to be able to easily reuse this elsewhere (and make sure it's working right).